### DISCUSSION TITLE
Kratos Simulation Versioning

### DISCUSSION THREAD

**Discussion by roigcarlo** [KratosMultiphysics/Kratos#11776] [General]
Dear @KratosMultiphysics/all 

Several of the teams working on Kratos have asked us to implement some kind of version control / tagging over the files generated to Kratos. In Plain words, to have them marked to know exactly for what version of Kratos were generated.

Piossible example (not final version):

`ProjectParameters.json`
```Json
{
    "problem_version": {
        "version_number": "9.4.1",
        "version_hash": "edcb5f605a",
        "version_branch": "master"
    }
    ...
}
```

After discussing in the @KratosMultiphysics/technical-committee we would like to know some insight about:
- Which files you are interested to have this version features.
- Should we keep the same version of Kratos or use a different one?
- Should files without version be considered to be files from the last version or from the first one?
- Any other suggestion that you may have

**Comment by loumalouomega** [comment]
In @KratosMultiphysics/altair we generate a ProjectParameters.json with the settings finally considered (with the defauls set when no defined), we could do that here and add the version that run an specific case.

**Comment by matekelemen** [comment]
Oh I like this idea a lot!

I'd also include information about
1) system that compiled Kratos (OS, its version, hardware - architecture, memory, GPU, etc.)
2) date of compilation
3) compiler (as well as its version, flags, and linker flags)
4) python (version, and all the information python provides about itself)

as well as information about the system Kratos is running on
- same info as `1` in the last list, but for the running system
- date of execution

I'd make this info available in the python layer, and stamp it in every output file that supports writing metadata. It would be a great help for tracking performance across PRs.

**Comment by matekelemen** [comment]
As for external libraries, I suggest copying the entire repo instead of just the important bits. I've already explained why this is necessary in #11795, but I'll write it down here as well for completeness' sake.

It'd make our lives easier when debugging, and would promote wider/easier adoption for companies that need to vet the software they're using. Having the entire external library available and providing exact version information about it would save the vetting team a lot of trouble, especially if the company is using the library for other software as well.

Furthermore, we should update the overview about the usage of our external libs (is it essential to Kratos; if not, link to instructions on how to disable it, etc...) in the main readme.

**Reply by loumalouomega** [reply]
I would be in favor of replacing the copy/paste stragey with a cmake madro which automatically downloads the last version, will be a compromise solution between the submodules and the manual copy. We can fork the libraries considered in order to ensure some stability (to avoid randomly breaking whenn a major release is done in the library side). We can even try even another alternatives, but basically I am in favor of removing the copy paste strategy. BTW, we have a dependecy manually patched that I should have PR in order to avoid future problems but I forgot :P (it is tinyexpr)

**Reply by matekelemen** [reply]
The problem with that approach is that
1) the dependencies are not clearly visible in the repository (they're tucked away somewhere in `CMakeLists.txt`)
2) spotting dependency version changes is more complicated (you'd need to check when that part of `CMakeLists.txt` was changed, **and** you need to check the dependency's repo)
3) downloading the sources would be a 2-stage process: first clone the Kratos repo, then clone the dependencies at Kratos build time. This is a minor issue, but unfortunately Germany is still in the stone age when it comes to internet access, and we need to plan accordingly.

**Reply by loumalouomega** [reply]
¯\_(ツ)_/¯

**Reply by roigcarlo** [reply]
I tend to partially agree with @matekelemen  for this particular issue. There are already mechanisms in cmake for doing that and its far from ideal. It can be used in a way were the version is explicitly mentioned and its visible. In fact if one wishes to, the version can be forwarded from the `configure.sh` or from a new `requirements.txt` akin to other projects. But in my opinion it is still very cumbersome for the final user to deal with.

For example, something I've been tempering with with the tests:

```CMake
# Testing
if(KRATOS_BUILD_TESTING MATCHES ON)   
  FetchContent_Declare(
    googletest
    URL https://github.com/google/googletest/archive/03597a01ee50ed33e9dfd640b249b4be3799d395.zip
  )
  # For Windows: Prevent overriding the parent project's compiler/linker settings
  set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
  FetchContent_MakeAvailable(googletest)
  enable_testing()
endif(KRATOS_BUILD_TESTING MATCHES ON) 
```

Here, `GTest` library is not uploaded to the Repo, so we avoid the first problem. The version is implicitly indicated inside the `CMakeList.txt`. If a version needs to be indicated from outside, something like this could be done:

```CMake
  FetchContent_Declare(
    googletest
    URL ${CMAKE_KRATOS_GTEST_VERSION}
  )
```

And then add it to the configure (pseudocode)

```bash
KRATOS_GTEST: {version: "X", url: "https://github.com/google/googletest/archive/03597a01ee50ed33e9dfd640b249b4be3799d395.zip"}

cmake ... -DCMAKE_KRATOS_GTEST_VERSION=${KRATOS_GTEST['URL']}
```

That presents two problems in my opinion:

- We cannot ensure the availability of the dependency 100% of the time and asses that everything is immutable upstream. This already has been a source of issues in dependencies working this way (most notably `MMG` which suddenly dropped the version we were using ). This is a shared problem with basically every solution relies on part of the code that is not owned by the repo.
 - All possible libraries that we are going to add using this mechanism need to be compatible with the mechanism per se. For example if `LIB_FOO_WHATEVER` is not prepared to be used using the `FetchContent_Declare` an ad-hoc replacement made by us needs to be in place. While its true that we can opt for the most generic retrieval mechanism possible, which basically end up being downloading the code manually and pasting it somewhere, which is what we are trying to avoid so its quite useless. Also we have to rely on the upstream and pray that they don't decide to change the distribution mechanism of the lib for whatever reason.

At the end, in my opinion all sums up to the fact that there is no centralized dependency manager in C++ (the intend of the 2nd code sniped) which is extended enough that we can make use of like `pip`, `npm` or `conda`/`mamba`. Maybe [Conan?](https://github.com/conan-io/conan), but still not all the dependencies we use are there. 

For the case of #11795, currently there is no other solution that copying all the lib as @mpentek says, since when we added `AMGCL` we did not foresee people wanting to use `VexCL` extensively (people doing it were basically the ones maintaining the lib in kratos). Something similar would happen if `pybind11` had other components. To be honest I have no clue on how to make it easier.

**Reply by loumalouomega** [reply]
Instead of Conan I recommend Spack, which is HPC focused.

**Reply by roigcarlo** [reply]
Yep well that is the problem, we want Spack for hpc, but we want something else for another user profile, see my point?

**Reply by loumalouomega** [reply]
:P
